File: LeetCode/binaryTree/654.maximum-binary-tree.py
# @before-stub-for-debug-begin
from python3problem654 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=654 lang=python3
#
# [654] Maximum Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> Optional[TreeNode]:
        if (not nums): return None

        # rootIdx = nums.index(max(nums)) # find argmax
        rootIdx = -1
        maxVal = -1
        for index, num in enumerate(nums):
            if num > maxVal:
                rootIdx = index
                maxVal = num

        leftTree = nums[0:rootIdx]
        rightTree = nums[rootIdx+1:]

        leftSubTree = self.constructMaximumBinaryTree(leftTree)
        rightSubTree = self.constructMaximumBinaryTree(rightTree)

        root = TreeNode(val=nums[rootIdx], left=leftSubTree, right=rightSubTree)

        return root   
# @lc code=end

File: LeetCode_Go/BinaryTree/654.maximum-binary-tree.go
/*
 * @lc app=leetcode id=654 lang=golang
 *
 * [654] Maximum Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Given an array of unique integers, construct a maximum binary tree using the following algorithm: create a root node with the maximum value in the array, then recursively build left and right subtrees from the subarrays to the left and right of the maximum value. Return the resulting maximum binary tree.

// follow the algorithm

// O(nlogn) time and O(n) space

func constructMaximumBinaryTree(nums []int) *TreeNode {
	root := constructMaximumBinaryTreeBetween(nums, 0, len(nums)-1)
	return root
}

func constructMaximumBinaryTreeBetween(nums []int, i, j int) *TreeNode {
	idx := findMaximumIdxBetween(nums, i, j)
	if idx == -1 {
		return nil
	}

	root := TreeNode{
		Val:   nums[idx],
		Left:  constructMaximumBinaryTreeBetween(nums, i, idx-1),
		Right: constructMaximumBinaryTreeBetween(nums, idx+1, j),
	}
	return &root
}

func findMaximumIdxBetween(nums []int, i, j int) int {
	if i > j {
		return -1
	}
	max := -1
	idx := -1
	for cur := i; cur <= j; cur++ {
		if nums[cur] > max {
			max = nums[cur]
			idx = cur
		}
	}
	return idx
}

// @lc code=end

